src/test/java/core/basesyntax/strategy/impl/PurchaseOperationHandlerTest.java

```diff
@@ -0,0 +1,81 @@
+package core.basesyntax.strategy.impl;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.FruitTransaction;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class PurchaseOperationHandlerTest {
+    private static final String BANANA = "banana";
+    private static final String APPLE = "apple";
+    private static final String NULL_MESSAGE = "Your FruitTransaction is Null";
+    private static final String NO_SUCH_FRUIT = "Your do not have such fruit";
+    private static PurchaseOperationHandler purchaseOperationHandler;
+
+    @BeforeAll
+    static void beforeAll() {
+        purchaseOperationHandler = new PurchaseOperationHandler();
+    }
+
+    @BeforeEach
+    void setUp() {
+        Storage.SHOPSTORAGE.clear();
+    }
+
+    @AfterEach
+    void tearDown() {
+        Storage.SHOPSTORAGE.clear();
+    }
+
+    @Test
+    void handle_PurchaseBanana_ok() {
+        Storage.SHOPSTORAGE.put(BANANA, 10);
+        FruitTransaction fruitPurchase = new FruitTransaction();
+        fruitPurchase.setFruit(BANANA);
+        fruitPurchase.setQuantity(5);
+        purchaseOperationHandler.handle(fruitPurchase);
+        assertEquals(5, Storage.SHOPSTORAGE.get(BANANA));
+    }
+
+    @Test
+    void handle_fruitNotInStorage_notOk() {
+        FruitTransaction noSuchFruit = new FruitTransaction();
+        noSuchFruit.setFruit(APPLE);
+        noSuchFruit.setQuantity(5);
+        assertThrows(RuntimeException.class,
+                () -> purchaseOperationHandler.handle(noSuchFruit),
+                NO_SUCH_FRUIT);
+    }
+
+    @Test
+    void handle_purchaseFruitTransactionNull_notOk() {
+        assertThrows(NullPointerException.class,

```

Comment:
```suggestion
        assertThrows(IllegalArgumentException.class,
```

Reply:
You should be using IllegalArgumentException (IAE), not NullPointerException (NPE) for the following reasons:

First, the [NPE JavaDoc](http://docs.oracle.com/javase/1.5.0/docs/api/java/lang/NullPointerException.html) explicitly lists the cases where NPE is appropriate. Notice that all of them are thrown by the runtime when null is used inappropriately. In contrast, the [IAE JavaDoc](http://docs.oracle.com/javase/1.5.0/docs/api/java/lang/IllegalArgumentException.html) couldn't be more clear: "Thrown to indicate that a method has been passed an illegal or inappropriate argument." Yup, that's you!

Second, when you see an NPE in a stack trace, what do you assume? Probably that someone dereferenced a null. When you see IAE, you assume the caller of the method at the top of the stack passed in an illegal value. Again, the latter assumption is true, the former is misleading.

Third, since IAE is clearly designed for validating parameters, you have to assume it as the default choice of exception, so why would you choose NPE instead? Certainly not for different behavior -- do you really expect calling code to catch NPE's separately from IAE and do something different as a result? Are you trying to communicate a more specific error message? But you can do that in the exception message text anyway, as you should for all other incorrect parameters.

Fourth, all other incorrect parameter data will be IAE, so why not be consistent? Why is it that an illegal null is so special that it deserves a separate exception from all other types of illegal arguments?

Finally, I accept the argument given by other answers that parts of the Java API use NPE in this manner. However, the Java API is inconsistent with everything from exception types to naming conventions, so I think just blindly copying (your favorite part of) the Java API isn't a good enough argument to trump these other considerations.

Reply:
Change it everywhere.

Reply:
To introduce a check for IllegalArgumentException, I need to modify the class itself in the FruitShop. I didn't include such checks there initially, but now I realize it was a mistake. I've decided to remove the checks for both NPE and IAE from the tests altogether.